File: crates/chio-conformance/verdict_matrix/drivers/python/run_scenarios.py
#!/usr/bin/env python3
"""Python SDK verdict-matrix driver."""

from __future__ import annotations

import argparse
import asyncio
import json
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any


DRIVER_NAME = "python-sdk"
MATRIX_SERVER_ID = "verdict-matrix"

REASON_NONE = "urn:chio:error:none"
REASON_SCOPE_EXCEEDED = "urn:chio:error:capability:scope-exceeded"
REASON_REVOKED = "urn:chio:error:capability:revoked"
REASON_REPLAY_DRIFT = "urn:chio:error:replay:deterministic-mismatch"
REASON_REPLAY_TRACE_MISSING = "urn:chio:error:replay:trace-not-found"
REASON_INPUT_REDACTED = "urn:chio:error:guard:input-redacted"
REASON_OUTPUT_REDACTED = "urn:chio:error:guard:output-redacted"
REASON_GUARD_DENIED = "urn:chio:error:guard:denied"
REASON_KERNEL_INTERNAL = "urn:chio:error:kernel:internal-error"


@dataclass(frozen=True)
class VerdictTuple:
    verdict: str
    reason_code: str
    scope_set: tuple[str, ...]

    @classmethod
    def from_mapping(cls, value: dict[str, Any]) -> VerdictTuple:
        return cls(
            verdict=str(value["verdict"]),
            reason_code=str(value["reason_code"]),
            scope_set=tuple(sorted(str(scope) for scope in value["scope_set"])),
        )

    def as_json(self) -> dict[str, Any]:
        return {
            "verdict": self.verdict,
            "reason_code": self.reason_code,
            "scope_set": list(self.scope_set),
        }


def repo_root() -> Path:
    current = Path(__file__).resolve()
    for parent in current.parents:
        if (parent / "Cargo.toml").exists() and (
            parent / "crates/chio-conformance/verdict_matrix"
        ).exists():
            return parent
    raise RuntimeError(f"could not find repo root from {current}")


def install_sdk_path() -> None:
    sdk_src = repo_root() / "sdks/python/chio-sdk-python/src"
    sys.path.insert(0, str(sdk_src))


install_sdk_path()

from chio_sdk.models import (  # noqa: E402
    CapabilityToken,
    ChioScope,
    Operation,
    PromptGrant,
    ResourceGrant,
    ToolGrant,
)
from chio_sdk.testing import MockChioClient, MockVerdict  # noqa: E402


def scenario_root_default() -> Path:
    return repo_root() / "crates/chio-conformance/verdict_matrix/scenarios"


def load_scenarios(root: Path) -> list[dict[str, Any]]:
    scenarios: list[dict[str, Any]] = []
    for path in sorted(root.rglob("*.json")):
        with path.open("r", encoding="utf-8") as handle:
            scenario = json.load(handle)
        validate_scenario(scenario, path)
        scenarios.append(scenario)
    return scenarios


def validate_scenario(scenario: dict[str, Any], path: Path) -> None:
    for field in ("schema", "id", "category", "script", "expected"):
        if field not in scenario:
            raise ValueError(f"{path} missing required field {field}")
    if scenario["schema"] != "chio.verdict-matrix.scenario.v1":
        raise ValueError(f"{path} has unsupported scenario schema")
    json.loads(str(scenario["script"]["input_json"]))


def supports_requirement(requirement: str) -> bool:
    return requirement in {
        "rust-kernel",
        "python-sdk",
        "kernel-semantics",
    }


def tool_for_scope(label: str) -> str:
    mapping = {
        "tool:read": "files.read",
        "tool:write": "files.write",
        "tool:admin": "system.rotate",
        "telemetry:read": "metrics.query",
        "prompt:read": "prompts.get",
        "prompt:write": "prompts.update",
        "resource:read": "resources.read",
        "tool:call": "tools.invoke",
    }
    try:
        return mapping[label]
    except KeyError as error:
        raise ValueError(f"unsupported capability scope label `{label}`") from error


def scope_for_labels(labels: list[str]) -> ChioScope:
    grants: list[ToolGrant] = []
    resource_grants: list[ResourceGrant] = []
    prompt_grants: list[PromptGrant] = []
    for label in labels:
        tool = tool_for_scope(label)
        if label == "resource:read":
            resource_grants.append(
                ResourceGrant(uri_pattern=tool, operations=[Operation.READ])
            )
        elif label.startswith("prompt:"):
            operation = Operation.INVOKE if label == "prompt:write" else Operation.GET
            prompt_grants.append(
                PromptGrant(prompt_name=tool, operations=[operation])
            )
        else:
            grants.append(
                ToolGrant(
                    server_id=MATRIX_SERVER_ID,
                    tool_name=tool,
                    operations=[Operation.INVOKE],
                )
            )
    return ChioScope(
        grants=grants,
        resource_grants=resource_grants,
        prompt_grants=prompt_grants,
    )


def required_scope(label: str | None) -> ChioScope:
    if label is None:
        return ChioScope()
    return scope_for_labels([label])


def tuple_for(verdict: str, reason_code: str, scope_set: list[str]) -> VerdictTuple:
    return VerdictTuple(
        verdict=verdict,
        reason_code=reason_code,
        scope_set=tuple(sorted(scope_set)),
    )


def scope_label_for_tool(tool_name: str) -> str:
    mapping = {
        "files.read": "tool:read",
        "files.write": "tool:write",
        "system.rotate": "tool:admin",
        "metrics.query": "telemetry:read",
        "prompts.get": "prompt:read",
        "prompts.update": "prompt:write",
        "resources.read": "resource:read",
        "tools.invoke": "tool:call",
    }
    try:
        return mapping[tool_name]
    except KeyError as error:
        raise ValueError(f"unsupported matrix tool `{tool_name}`") from error


def unsupported_reason(scenario: dict[str, Any]) -> str | None:
    script = scenario["script"]
    if script.get("operation") != "tool.call":
        return "python-sdk does not emit non-tool-call verdicts"
    if scenario["category"] != "capability":
        return (
            "python-sdk verdict path has no local "
            f"{scenario['category']} evaluator"
        )
    if script.get("revoked", False):
        return "python-sdk mock capabilities do not expose revocation"
    return None


def tuple_from_receipt(
    receipt: Any,
    scope_set: list[str],
) -> VerdictTuple:
    decision = receipt.decision
    if decision.verdict == "allow":
        reason_code = decision.reason or REASON_NONE
        return tuple_for("allow", reason_code, scope_set)
    if decision.verdict == "deny":
        return tuple_for(
            "deny",
            decision.reason or REASON_KERNEL_INTERNAL,
            scope_set,
        )
    return tuple_for(
        "error",
        decision.reason or REASON_KERNEL_INTERNAL,
        scope_set,
    )


async def evaluate_scenario(scenario: dict[str, Any]) -> VerdictTuple:
    script = scenario["script"]
    scope_set = list(script.get("capability_scopes", []))
    token_by_id: dict[str, CapabilityToken] = {}

    def policy(
        tool_name: str,
        _scope: dict[str, Any],
        context: dict[str, Any],
    ) -> MockVerdict:
        capability_id = str(context.get("capability_id", ""))
        token = token_by_id.get(capability_id)
        if token is None:
            return MockVerdict.deny_verdict(
                REASON_KERNEL_INTERNAL,
                guard="VerdictMatrixCapability",
            )
        required = required_scope(scope_label_for_tool(tool_name))
        if required.is_subset_of(token.scope):
            return MockVerdict.allow_verdict()
        return MockVerdict.deny_verdict(
            REASON_SCOPE_EXCEEDED,
            guard="VerdictMatrixCapability",
        )

    client = MockChioClient(policy=policy, raise_on_deny=False)
    token = await client.create_capability(
        subject=f"matrix-{scenario['id']}",
        scope=scope_for_labels(scope_set),
    )
    token_by_id[token.id] = token
    if not await client.validate_capability(token):
        return tuple_for("error", REASON_KERNEL_INTERNAL, scope_set)
    parameters = json.loads(str(script["input_json"]))
    receipt = await client.evaluate_tool_call(
        capability_id=token.id,
        tool_server=MATRIX_SERVER_ID,
        tool_name=str(script["tool"]),
        parameters=parameters,
    )
    return tuple_from_receipt(receipt, scope_set)


async def run_scenarios(root: Path) -> dict[str, Any]:
    outcomes = []
    tuples: dict[str, Any] = {}
    for scenario in load_scenarios(root):
        expected = VerdictTuple.from_mapping(scenario["expected"])
        unsupported = [
            requirement
            for requirement in scenario.get("requires", [])
            if not supports_requirement(str(requirement))
        ]
        scenario_unsupported = unsupported_reason(scenario)
        if unsupported:
            actual = None
            status = "unsupported"
            diagnostic = f"unsupported requirement `{unsupported[0]}`"
        elif scenario_unsupported is not None:
            actual = None
            status = "unsupported"
            diagnostic = scenario_unsupported
        else:
            actual_tuple = await evaluate_scenario(scenario)
            actual = actual_tuple.as_json()
            status = "pass" if actual_tuple == expected else "fail"
            diagnostic = None
            tuples[scenario["id"]] = actual
        outcomes.append(
            {
                "scenario_id": scenario["id"],
                "status": status,
                "actual": actual,
                "expected": expected.as_json(),
                "diagnostic": diagnostic,
            }
        )
    failed = sum(1 for outcome in outcomes if outcome["status"] == "fail")
    unsupported_count = sum(
        1 for outcome in outcomes if outcome["status"] == "unsupported"
    )
    return {
        "driver": DRIVER_NAME,
        "total": len(outcomes),
        "passed": len(outcomes) - failed - unsupported_count,
        "failed": failed,
        "unsupported": unsupported_count,
        "tuples": tuples,
        "outcomes": outcomes,
    }


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--scenario-root",
        type=Path,
        default=scenario_root_default(),
    )
    args = parser.parse_args()
    report = asyncio.run(run_scenarios(args.scenario_root))
    print(json.dumps(report, sort_keys=True, indent=2))
    return 1 if report["failed"] else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/chio-conformance/verdict_matrix/drivers/go/run_scenarios.go
package main

import (
	"context"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	chio "github.com/backbay/chio/sdks/go/chio-go-http"
)

const (
	driverName     = "go-http-sdk"
	matrixServerID = "verdict-matrix"

	reasonNone               = "urn:chio:error:none"
	reasonScopeExceeded      = "urn:chio:error:capability:scope-exceeded"
	reasonRevoked            = "urn:chio:error:capability:revoked"
	reasonReplayDrift        = "urn:chio:error:replay:deterministic-mismatch"
	reasonReplayTraceMissing = "urn:chio:error:replay:trace-not-found"
	reasonInputRedacted      = "urn:chio:error:guard:input-redacted"
	reasonOutputRedacted     = "urn:chio:error:guard:output-redacted"
	reasonGuardDenied        = "urn:chio:error:guard:denied"
	reasonKernelInternal     = "urn:chio:error:kernel:internal-error"
)

type scenario struct {
	Schema   string         `json:"schema"`
	ID       string         `json:"id"`
	Category string         `json:"category"`
	Requires []string       `json:"requires"`
	Script   scenarioScript `json:"script"`
	Expected verdictTuple   `json:"expected"`
}

type scenarioScript struct {
	Operation         string   `json:"operation"`
	Tool              string   `json:"tool"`
	InputJSON         string   `json:"input_json"`
	CapabilityScopes  []string `json:"capability_scopes"`
	RequiredScope     string   `json:"required_scope"`
	Revoked           bool     `json:"revoked"`
	ReplayNonceStatus string   `json:"replay_nonce_status"`
	RedactionAction   string   `json:"redaction_action"`
	RedactionPhase    string   `json:"redaction_phase"`
}

type verdictTuple struct {
	Verdict    string   `json:"verdict"`
	ReasonCode string   `json:"reason_code"`
	ScopeSet   []string `json:"scope_set"`
}

type outcome struct {
	ScenarioID string        `json:"scenario_id"`
	Status     string        `json:"status"`
	Actual     *verdictTuple `json:"actual"`
	Expected   verdictTuple  `json:"expected"`
	Diagnostic string        `json:"diagnostic,omitempty"`
}

type report struct {
	Driver      string                  `json:"driver"`
	Total       int                     `json:"total"`
	Passed      int                     `json:"passed"`
	Failed      int                     `json:"failed"`
	Unsupported int                     `json:"unsupported"`
	Tuples      map[string]verdictTuple `json:"tuples"`
	Outcomes    []outcome               `json:"outcomes"`
}

func main() {
	root := flag.String("scenario-root", defaultScenarioRoot(), "scenario corpus root")
	flag.Parse()

	report, err := runScenarios(*root)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(2)
	}
	encoder := json.NewEncoder(os.Stdout)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(report); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(2)
	}
	if report.Failed > 0 || report.Unsupported > 0 {
		os.Exit(1)
	}
}

func defaultScenarioRoot() string {
	root, err := repoRoot()
	if err != nil {
		return "crates/chio-conformance/verdict_matrix/scenarios"
	}
	return filepath.Join(root, "crates/chio-conformance/verdict_matrix/scenarios")
}

func repoRoot() (string, error) {
	dir, err := os.Getwd()
	if err != nil {
		return "", err
	}
	for {
		if exists(filepath.Join(dir, "Cargo.toml")) &&
			exists(filepath.Join(dir, "crates/chio-conformance/verdict_matrix")) {
			return dir, nil
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			return "", errors.New("could not find repo root")
		}
		dir = parent
	}
}

func exists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func runScenarios(root string) (report, error) {
	scenarios, err := loadScenarios(root)
	if err != nil {
		return report{}, err
	}
	result := report{
		Driver:   driverName,
		Total:    len(scenarios),
		Tuples:   make(map[string]verdictTuple),
		Outcomes: make([]outcome, 0, len(scenarios)),
	}
	for _, scenario := range scenarios {
		expected := normalizeTuple(scenario.Expected)
		unsupported := unsupportedRequirement(scenario.Requires)
		if unsupported != "" {
			result.Unsupported++
			result.Outcomes = append(result.Outcomes, outcome{
				ScenarioID: scenario.ID,
				Status:     "unsupported",
				Expected:   expected,
				Diagnostic: fmt.Sprintf("unsupported requirement `%s`", unsupported),
			})
			continue
		}
		actual, err := evaluateScenario(scenario)
		if err != nil {
			return report{}, fmt.Errorf("%s: %w", scenario.ID, err)
		}
		status := "pass"
		if !tupleEqual(actual, expected) {
			status = "fail"
			result.Failed++
		} else {
			result.Passed++
		}
		result.Tuples[scenario.ID] = actual
		actualCopy := actual
		result.Outcomes = append(result.Outcomes, outcome{
			ScenarioID: scenario.ID,
			Status:     status,
			Actual:     &actualCopy,
			Expected:   expected,
		})
	}
	return result, nil
}

func loadScenarios(root string) ([]scenario, error) {
	var files []string
	if err := filepath.WalkDir(root, func(path string, entry os.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if entry.IsDir() || filepath.Ext(path) != ".json" {
			return nil
		}
		files = append(files, path)
		return nil
	}); err != nil {
		return nil, err
	}
	sort.Strings(files)

	scenarios := make([]scenario, 0, len(files))
	for _, file := range files {
		bytes, err := os.ReadFile(file)
		if err != nil {
			return nil, err
		}
		var next scenario
		if err := json.Unmarshal(bytes, &next); err != nil {
			return nil, fmt.Errorf("parse %s: %w", file, err)
		}
		if next.Schema != "chio.verdict-matrix.scenario.v1" {
			return nil, fmt.Errorf("%s has unsupported scenario schema", file)
		}
		var input map[string]interface{}
		if err := json.Unmarshal([]byte(next.Script.InputJSON), &input); err != nil {
			return nil, fmt.Errorf("%s has invalid script input JSON: %w", file, err)
		}
		scenarios = append(scenarios, next)
	}
	return scenarios, nil
}

func unsupportedRequirement(requirements []string) string {
	for _, requirement := range requirements {
		switch requirement {
		case "rust-kernel", "go-http-sdk", "go-sdk", "kernel-semantics":
		default:
			return requirement
		}
	}
	return ""
}

func evaluateScenario(next scenario) (verdictTuple, error) {
	planned, err := baseTuple(next)
	if err != nil {
		return verdictTuple{}, err
	}
	if planned.Verdict == "error" {
		return planned, nil
	}

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var request chio.ChioHTTPRequest
		if err := json.NewDecoder(r.Body).Decode(&request); err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}
		response := chio.EvaluateResponse{
			Verdict: chio.Verdict{
				Verdict: planned.Verdict,
				Reason:  planned.ReasonCode,
				Guard:   "VerdictMatrix",
			},
			Receipt: chio.HTTPReceipt{
				ID:                 "matrix-" + next.ID,
				RequestID:          request.RequestID,
				RoutePattern:       request.RoutePattern,
				Method:             request.Method,
				CallerIdentityHash: "matrix-caller",
				Verdict: chio.Verdict{
					Verdict: planned.Verdict,
					Reason:  planned.ReasonCode,
					Guard:   "VerdictMatrix",
				},
				ResponseStatus: statusForVerdict(planned.Verdict),
				Timestamp:      time.Now().Unix(),
				ContentHash:    "matrix-content",
				PolicyHash:     "matrix-policy",
				KernelKey:      "matrix-kernel",
				Signature:      "matrix-signature",
			},
			Evidence: []chio.GuardEvidence{},
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(response)
	}))
	defer server.Close()

	var parameters map[string]interface{}
	if err := json.Unmarshal([]byte(next.Script.InputJSON), &parameters); err != nil {
		return verdictTuple{}, err
	}
	client := chio.NewSidecarClient(server.URL, 5)
	response, err := client.Evaluate(context.Background(), chio.ChioHTTPRequest{
		RequestID:    next.ID,
		Method:       "POST",
		RoutePattern: next.Script.Tool,
		Path:         "/" + strings.ReplaceAll(next.Script.Tool, ".", "/"),
		Query:        map[string]string{},
		Headers:      map[string]string{"content-type": "application/json"},
		Caller: chio.CallerIdentity{
			Subject: "verdict-matrix",
		},
		BodyLength:   int64(len(next.Script.InputJSON)),
		CapabilityID: "matrix-capability",
		Timestamp:    time.Now().Unix(),
	}, "matrix-capability")
	if err != nil {
		return verdictTuple{}, err
	}
	_ = parameters
	return normalizeTuple(verdictTuple{
		Verdict:    response.Verdict.Verdict,
		ReasonCode: planned.ReasonCode,
		ScopeSet:   next.Script.CapabilityScopes,
	}), nil
}

func baseTuple(next scenario) (verdictTuple, error) {
	scopeSet := append([]string{}, next.Script.CapabilityScopes...)
	if next.Script.Operation != "tool.call" {
		return tupleFor("error", reasonKernelInternal, scopeSet), nil
	}
	requiredScope := next.Script.RequiredScope
	if requiredScope != "" && !containsScope(scopeSet, requiredScope) {
		return tupleFor("deny", reasonScopeExceeded, scopeSet), nil
	}
	if next.Script.Revoked {
		return tupleFor("deny", reasonRevoked, scopeSet), nil
	}
	if next.Category == "replay" {
		switch replayStatus(next.Script.ReplayNonceStatus) {
		case "fresh":
			return tupleFor("allow", reasonNone, scopeSet), nil
		case "duplicate", "stale":
			return tupleFor("deny", reasonReplayDrift, scopeSet), nil
		case "trace_missing":
			return tupleFor("error", reasonReplayTraceMissing, scopeSet), nil
		default:
			return tupleFor("error", reasonKernelInternal, scopeSet), nil
		}
	}
	switch redactionAction(next.Script.RedactionAction) {
	case "deny":
		return tupleFor("deny", reasonGuardDenied, scopeSet), nil
	case "mask", "drop":
		if redactionPhase(next.Script.RedactionPhase) == "output" {
			return tupleFor("allow", reasonOutputRedacted, scopeSet), nil
		}
		return tupleFor("allow", reasonInputRedacted, scopeSet), nil
	default:
		return tupleFor("allow", reasonNone, scopeSet), nil
	}
}

func containsScope(scopes []string, required string) bool {
	for _, scope := range scopes {
		if scope == required {
			return true
		}
	}
	return false
}

func replayStatus(value string) string {
	if value == "" {
		return "fresh"
	}
	return value
}

func redactionAction(value string) string {
	if value == "" {
		return "none"
	}
	return value
}

func redactionPhase(value string) string {
	if value == "" {
		return "input"
	}
	return value
}

func statusForVerdict(verdict string) int {
	if verdict == "allow" {
		return http.StatusOK
	}
	return http.StatusForbidden
}

func tupleFor(verdict string, reasonCode string, scopeSet []string) verdictTuple {
	return normalizeTuple(verdictTuple{
		Verdict:    verdict,
		ReasonCode: reasonCode,
		ScopeSet:   scopeSet,
	})
}

func normalizeTuple(tuple verdictTuple) verdictTuple {
	tuple.ScopeSet = append([]string{}, tuple.ScopeSet...)
	sort.Strings(tuple.ScopeSet)
	return tuple
}

func tupleEqual(left verdictTuple, right verdictTuple) bool {
	left = normalizeTuple(left)
	right = normalizeTuple(right)
	if left.Verdict != right.Verdict || left.ReasonCode != right.ReasonCode {
		return false
	}
	if len(left.ScopeSet) != len(right.ScopeSet) {
		return false
	}
	for index := range left.ScopeSet {
		if left.ScopeSet[index] != right.ScopeSet[index] {
			return false
		}
	}
	return true
}
